terraform-provider-constellation/internal/provider/cluster_resource.go

```diff
@@ -149,155 +314,569 @@ func (r *ClusterResource) Configure(_ context.Context, req resource.ConfigureReq
-	//     return
-	// }
+	// All Calls to the Constellation API are idempotent, thus we don't need to implement reading.
+
+	// Alternatively, we could:
+
+	// Retrieve more up-to-date data from the cluster. e.g.:
+	// - CSI Driver enabled?
+	// - Kubernetes version?
+	// - Microservice version?
+	// - Attestation Config?
 
 	// Save updated data into Terraform state
 	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
 }
 
 // Update is called when the resource is updated.
 func (r *ClusterResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
-	var data ClusterResourceModel
-
 	// Read Terraform plan data into the model
+	var data ClusterResourceModel
 	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
-
 	if resp.Diagnostics.HasError() {
 		return
 	}
 
-	// If applicable, this is a great opportunity to initialize any necessary
-	// provider client data and make a call using it.
-	// httpResp, err := r.client.Do(httpReq)
-	// if err != nil {
-	//     resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to update example, got error: %s", err))
-	//     return
-	// }
+	// Apply changes to the cluster, skipping the init RPC.
+	diags := r.apply(ctx, data, true, false)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
 	// Save updated data into Terraform state
 	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
 }
 
 // Delete is called when the resource is destroyed.
 func (r *ClusterResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
-	var data ClusterResourceModel
-
 	// Read Terraform prior state data into the model
+	var data ClusterResourceModel
 	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
-
 	if resp.Diagnostics.HasError() {
 		return
 	}
-
-	// If applicable, this is a great opportunity to initialize any necessary
-	// provider client data and make a call using it.
-	// httpResp, err := r.client.Do(httpReq)
-	// if err != nil {
-	//     resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to delete example, got error: %s", err))
-	//     return
-	// }
 }
 
 // ImportState imports to the resource.
-func (r *ClusterResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
-	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
+func (r *ClusterResource) ImportState(_ context.Context, _ resource.ImportStateRequest, _ *resource.ImportStateResponse) {
+	// TODO: Implement
+
+	// Take Kubeconfig, Cluster Endpoint and Master Secret and save to state
 }
 
-type tfLogger struct {
-	dg *diag.Diagnostics
+// apply applies changes to a cluster. It can be used for both creating and updating a cluster.
+// This implements the core part of the Create and Update methods.
+func (r *ClusterResource) apply(ctx context.Context, data ClusterResourceModel, skipInitRPC, skipNodeUpgrade bool) diag.Diagnostics {
```

Comment:
nit: I think the init RC should not be part of apply since it is only needed for create. also removes a bool flag.

Reply:
But it would also require us to add a lot of inputs to the apply function, since they need to be available within the function as well as to the RPC. I think it's best to leave as is for now.